// dharma-js/src/Servicing.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _bignumber = require('bignumber.js');

var _bignumber2 = _interopRequireDefault(_bignumber);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _Constants = require('./Constants.js');

var _Constants2 = _interopRequireDefault(_Constants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Servicing = function () {
  function Servicing(loan) {
    _classCallCheck(this, Servicing);

    this.loan = loan;
  }

  _createClass(Servicing, [{
    key: 'expectedAmountRepaidByDate',
    value: function expectedAmountRepaidByDate(date) {
      // if (this.loan.state !== Constants.ACCEPTED_STATE)
      //   throw new Error('Loan must be in ACCEPTED state before servicing ' +
      //     'utilities can be accessed');

      var expectedPeriodicRepayment = this.periodicRepaymentOwed();

      var numRepaymentPeriods = this._numRepaymentPeriods(date);

      return expectedPeriodicRepayment.times(numRepaymentPeriods);
    }
  }, {
    key: 'getInterestRepaidToDate',
    value: function () {
      var _ref = _asyncToGenerator(regeneratorRuntime.mark(function _callee() {
        var amountRepaid, decimals, interestRate;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.loan.amountRepaid();

              case 2:
                amountRepaid = _context.sent;
                decimals = Math.pow(10, 18);
                interestRate = this.loan.interestRate;

                interestRate = interestRate.div(decimals);

                return _context.abrupt('return', amountRepaid.times(interestRate).div(interestRate.plus(1)));

              case 7:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getInterestRepaidToDate() {
        return _ref.apply(this, arguments);
      }

      return getInterestRepaidToDate;
    }()
  }, {
    key: 'getPrincipalRepaidToDate',
    value: function () {
      var _ref2 = _asyncToGenerator(regeneratorRuntime.mark(function _callee2() {
        var amountRepaid, interestRepaidToDate;
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.loan.amountRepaid();

              case 2:
                amountRepaid = _context2.sent;
                _context2.next = 5;
                return this.getInterestRepaidToDate();

              case 5:
                interestRepaidToDate = _context2.sent;
                return _context2.abrupt('return', amountRepaid.minus(interestRepaidToDate));

              case 7:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getPrincipalRepaidToDate() {
        return _ref2.apply(this, arguments);
      }

      return getPrincipalRepaidToDate;
    }()
  }, {
    key: '_numRepaymentPeriods',
    value: function _numRepaymentPeriods(date) {
      var termBeginDate = new Date(this.loan.termBeginTimestamp * 1000);
      var numPeriods = this._numPeriodsBetween(termBeginDate, date, this.loan.terms.periodType(), this.loan.terms.periodLength());
      return _bignumber2.default.min(numPeriods, this.loan.terms.termLength());
    }
  }, {
    key: 'getRepaymentStatus',
    value: function () {
      var _ref3 = _asyncToGenerator(regeneratorRuntime.mark(function _callee3() {
        var amountRepaid, expectedAmountRepaid, durationSinceTermBegin, numPeriodsRepaid, lastRepaymentDateMissed, weeksSinceRepaymentDateMissed;
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.loan.amountRepaid();

              case 2:
                amountRepaid = _context3.sent;
                expectedAmountRepaid = this.expectedAmountRepaidByDate(new Date());

                if (!amountRepaid.gte(expectedAmountRepaid)) {
                  _context3.next = 13;
                  break;
                }

                durationSinceTermBegin = _moment2.default.duration((0, _moment2.default)().diff(this.loan.termBeginTimestamp * 1000));

                if (!(durationSinceTermBegin < this.termDuration())) {
                  _context3.next = 10;
                  break;
                }

                return _context3.abrupt('return', 'CURRENT');

              case 10:
                return _context3.abrupt('return', 'REPAID');

              case 11:
                _context3.next = 21;
                break;

              case 13:
                numPeriodsRepaid = amountRepaid.div(this.periodicRepaymentOwed()).floor().toNumber();
                lastRepaymentDateMissed = (0, _moment2.default)(this.loan.termBeginTimestamp * 1000).add(this.periodDuration(numPeriodsRepaid + 1));
                weeksSinceRepaymentDateMissed = _moment2.default.duration((0, _moment2.default)().diff(lastRepaymentDateMissed)).asWeeks();

                if (!(weeksSinceRepaymentDateMissed <= 2)) {
                  _context3.next = 20;
                  break;
                }

                return _context3.abrupt('return', 'DELINQUENT');

              case 20:
                return _context3.abrupt('return', 'DEFAULT');

              case 21:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getRepaymentStatus() {
        return _ref3.apply(this, arguments);
      }

      return getRepaymentStatus;
    }()
  }, {
    key: 'getRepaymentDates',
    value: function getRepaymentDates() {
      var dates = [];
      var termBeginDate = new Date(this.loan.termBeginTimestamp * 1000);
      for (var i = 1; i <= this.loan.terms.termLength(); i++) {
        var repaymentDate = (0, _moment2.default)(termBeginDate).add(this.periodDuration(i)).toDate();
        dates.push(repaymentDate);
      }
      return dates;
    }
  }, {
    key: 'periodicRepaymentOwed',
    value: function periodicRepaymentOwed() {
      return this.totalOwed().div(this.loan.terms.termLength());
    }
  }, {
    key: 'totalOwed',
    value: function totalOwed() {
      var decimals = Math.pow(10, 18);

      var interestRate = this.loan.interestRate;
      interestRate = interestRate.div(decimals).plus(1);

      var principal = this.loan.principal;
      return interestRate.times(principal);
    }
  }, {
    key: 'currentBalanceOwed',
    value: function () {
      var _ref4 = _asyncToGenerator(regeneratorRuntime.mark(function _callee4() {
        var decimals, amountRepaid, expectedAmountRepaid;
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                decimals = Math.pow(10, 18);
                _context4.next = 3;
                return this.loan.amountRepaid();

              case 3:
                amountRepaid = _context4.sent;
                expectedAmountRepaid = this.expectedAmountRepaidByDate(new Date());
                return _context4.abrupt('return', _bignumber2.default.max(expectedAmountRepaid.minus(amountRepaid), 0));

              case 6:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function currentBalanceOwed() {
        return _ref4.apply(this, arguments);
      }

      return currentBalanceOwed;
    }()
  }, {
    key: 'periodDuration',
    value: function periodDuration() {
      var numPeriods = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;

      switch (this.loan.terms.periodType()) {
        case 'daily':
          return _moment2.default.duration(numPeriods * this.loan.terms.periodLength(), 'days');
          break;
        case 'weekly':
          return _moment2.default.duration(numPeriods * this.loan.terms.periodLength(), 'weeks');
          break;
        case 'monthly':
          return _moment2.default.duration(numPeriods * this.loan.terms.periodLength(), 'months');
          break;
        case 'yearly':
          return _moment2.default.duration(numPeriods * this.loan.terms.periodLength(), 'years');
          break;
      }
    }
  }, {
    key: 'termDuration',
    value: function termDuration() {
      return this.periodDuration(this.loan.terms.termLength());
    }
  }, {
    key: '_numPeriodsBetween',
    value: function _numPeriodsBetween(startDate, endDate, periodType, periodLength) {
      var startDateWrapper = (0, _moment2.default)(startDate);
      var endDateWrapper = (0, _moment2.default)(endDate);

      var timeDiff = _moment2.default.duration(endDateWrapper.diff(startDateWrapper));

      switch (periodType) {
        case 'daily':
          return Math.floor(timeDiff.asDays() / periodLength);
          break;
        case 'weekly':
          return Math.floor(timeDiff.asWeeks() / periodLength);
          break;
        case 'monthly':
          return Math.floor(timeDiff.asMonths() / periodLength);
          break;
        case 'yearly':
          return Math.floor(timeDiff.asYears() / periodLength);
          break;
      }
    }
  }]);

  return Servicing;
}();

module.exports = Servicing;
